File: Software/workspace/SimHughBot/src/Commands/TurnToAngle.cpp
#include "TurnToAngle.h"


#define P 0.1
#define I 0.0
#define D 0.1

#define WIDTH 25 // horizontal distance between wheels (side-to-side)
#define LENGTH 8 // vertical distance between center wheels only (i.e the wheels with encoders)

#define USE_GYRO
#define DEBUG_COMMAND

TurnToAngle::TurnToAngle(double a) : CommandBase("DriveStraight"),
	pid(P,I,D,this,this)
{
	Requires(driveTrain.get());
	angle = a;
	// radius of travel circle = 1/2 diagonal of rectangle containing center wheels
  	radius = 0.5*sqrt(WIDTH*WIDTH+LENGTH*LENGTH);
	std::cout << "new TurnToAngle("<<a<<")"<< std::endl;
}

// Called just before this Command runs the first time
void TurnToAngle::Initialize() {
#ifdef USE_GYRO
	double a=angle;
#else
	double d = driveTrain->GetDistance();
	driveTrain->Reset();
	driveTrain->SetDistance(d);
	double a = angle*2.0*M_PI*radius/360;
#endif
  	pid.Reset();
  	// arc length (wheel travel distance) given radius and turn angle
  	pid.SetSetpoint(a);
	pid.SetAbsoluteTolerance(1);
	pid.Enable();
	driveTrain->Enable();
	std::cout << "TurnToAngle Started: "<< a <<std::endl;
}

// Called repeatedly when this Command is scheduled to run
void TurnToAngle::Execute() {
   // nothing to do: action is in PID control loop
}

// Make this return true when this Command no longer needs to run execute()
bool TurnToAngle::IsFinished() {
	return pid.OnTarget();
}

// Called once after isFinished returns true
void TurnToAngle::End() {
	pid.Disable();
    std::cout << "TurnToAngle End" << std::endl;
}

// Called when another command which requires one or more of the same
// subsystems is scheduled to run
void TurnToAngle::Interrupted() {
	End();
}
double TurnToAngle::PIDGet() {
#ifdef USE_GYRO
	double a=driveTrain->GetHeading();
#ifdef DEBUG_COMMAND
	std::cout << "TurnToAngle::PIDGet:" << a << std::endl;
#endif

	return driveTrain->GetHeading();
#else
	double l = driveTrain->GetLeftDistance();
	double r = driveTrain->GetRightDistance();
	double d = 0.5 * (fabs(l) + fabs(r)); // average wheel travel distance

	d = angle < 0 ? -d : d;
	double a = d * 360 / M_PI / radius / 2;

	driveTrain->SetAngle(a);
#ifdef DEBUG_COMMAND
	std::cout << "TurnToAngle::PIDGet:" << l << "," << r << ","<<d<<","<<a << std::endl;
#endif
	return d;
#endif
}

void TurnToAngle::PIDWrite(double a) {
    std::cout << "TurnToAngle::PIDWrite("<<a<<")"<< std::endl;
	driveTrain->TankDrive(a,-a);

}

File: Software/workspace/JSim2018Bot/src/org/usfirst/frc/team159/robot/commands/TurnToAngle.java
package org.usfirst.frc.team159.robot.commands;

import org.usfirst.frc.team159.robot.Robot;

import edu.wpi.first.wpilibj.PIDController;
import edu.wpi.first.wpilibj.PIDOutput;
import edu.wpi.first.wpilibj.PIDSource;
import edu.wpi.first.wpilibj.PIDSourceType;
import edu.wpi.first.wpilibj.command.TimedCommand;

/**
 *
 */
public class TurnToAngle extends TimedCommand implements PIDSource, PIDOutput{
    double angle;
    PIDController pid;
    PIDSourceType type = PIDSourceType.kDisplacement;

    public TurnToAngle(double angle,double tm) {
        super(tm);
        requires(Robot.driveTrain);
        this.angle=angle;
        pid = new PIDController(0.005, 0.0, 0.0, this, this, 0.01);
    }

    // Called just before this Command runs the first time
    protected void initialize() {
      pid.reset();
      pid.setSetpoint(angle);
      pid.setAbsoluteTolerance(4.0);
      pid.enable();
      System.out.println("TurnToAngle::initialize:"+angle);
    }

    // Called repeatedly when this Command is scheduled to run
    protected void execute() {
    }
    protected boolean isFinished() {
      if (super.isFinished()|| pid.onTarget())
        return true;
      return false;
    }

    // Called once after timeout
    protected void end() {
      System.out.println("TurnToAngle::end()");
      pid.disable();
    }

    // Called when another command which requires one or more of the same
    // subsystems is scheduled to run
    protected void interrupted() {
      System.out.println("TurnToAngle::interrupted()");
      end();
    }

    @Override
    public void pidWrite(double d) {
      Robot.driveTrain.set(-d, d);
    }

    @Override
    public void setPIDSourceType(PIDSourceType pidSource) {
      type = pidSource;      
    }

    @Override
    public PIDSourceType getPIDSourceType() {
      return type;
    }

    @Override
    public double pidGet() {
      return Robot.driveTrain.getHeading();

    }
}
